### PR TITLE
Improve network-status processing

### PR DESCRIPTION
**Pull request by orelmisan**
### What this PR does
The value of the pod's `k8s.v1.cni.cncf.io/network-status` annotation is used for:
1. Updating the VMI's interfaces status.
2. Calculating the value of the `kubevirt.io/network-info` annotation.
3. Creating an ordinal naming scheme for a migration target pod in case the source pod uses it.
4. Discovering whether the VMI uses an ordinal naming scheme in VM Controller.

This PR:
1. Centralizes the logic responsible for parsing the value of the `k8s.v1.cni.cncf.io/network-status` annotation under `pkg/network/multus`.
2. Uses processed Go structs where applicable instead of raw strings.
3. Removes unnecessary mapping of NetworkStatus by pod interface name.
4. Cleans up after #13018.
 

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [X] PR: The PR description is expressive enough and will help future contributors
- [X] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [X] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [X] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/network/deviceinfo/deviceinfo_test.go
Comment (orelmisan) [reply pkg/network/deviceinfo/deviceinfo_test.go:101]: In this case, two maps are compared for equality.
Gomaga's Equal function can compare two maps for equality.
Please see [1][2].

BTW, the assertion haven't changed by the PR (other than the function's inputs and the removal of newlines).

[1] https://onsi.github.io/gomega/#equalexpected-interface
[2] https://pkg.go.dev/reflect#DeepEqual

File: pkg/network/pod/annotations/generator.go
Comment (orelmisan) [reply pkg/network/pod/annotations/generator.go:117]: Prior to this change, `MapNetworkNameToDeviceInfo` could have returned an error in one of the following cases:
1. The value of the network-status annotation was empty.
2. The unmarshaling of the raw JSON string to `[]networkv1.NetworkStatus` had failed for some reason.

By making `MapNetworkNameToDeviceInfo` take a `[]networkv1.NetworkStatus` as an input parameter - the two possible error paths are no longer relevant - thus there is no need for this warning IMO.

Please note that `multus.NetworkStatusesFromPod` has a log message in case the JSON unmarshaling fails:
https://github.com/kubevirt/kubevirt/blob/635ba09c13144b3deecd1114c88ec50bc9d65dd4/pkg/network/multus/status.go#L45-L55

File: pkg/network/multus/status.go
Comment (orelmisan) [reply pkg/network/multus/status.go:32]: It actually maps the entire slice (unlike `NonDefaultNetworkStatusIndexedByPodIfaceName`), please see the unit tests.

File: pkg/network/multus/status.go
Comment (orelmisan) [reply pkg/network/multus/status.go:33]: According to [Kubernetes Network Custom Resource Definition De-facto Standard](https://github.com/k8snetworkplumbingwg/multi-net-spec/blob/master/v1.3/%5Bv1.3%5D%20Kubernetes%20Network%20Custom%20Resource%20Definition%20De-facto%20Standard.pdf):

> 4. Network Attachment Selection Annotation
To select one or more secondary ("sidecar") networks to which a pod should be attached, this specification defines a Pod object annotation

When using secondary networks (or binding plugins with CNI) in KubeVirt, we use the `k8s.v1.cni.cncf.io/networks` annotation to connect to these networks.

The value of `k8s.v1.cni.cncf.io/network-status` represents the successful network attachments - meaning that multiple network attachments could be represented by it.
This is the reason it unmarshals to a `[]networkv1.NetworkStatus` - one element per network.

One could look at the entire slice and consider it as the overall "pod network status", but a different approach could consider each element as a single network status.

File: pkg/virt-controller/watch/vmi/vmi.go
Comment (orelmisan) [reply pkg/virt-controller/watch/vmi/vmi.go:2179]: Is there a scenario where it is possible for its `networkNameScheme` parameter to be empty?

Currently, it is called from:
1. `GenerateCNIAnnotation` with hashed naming scheme.
2. `Generator.GenerateFromSource` with ordinal naming scheme.
3. `Controller.updateMultusAnnotation` with hashed or ordinal naming scheme.

BTW, it is not covered by unit tests.

File: pkg/network/namescheme/networknamescheme.go
Comment (orelmisan) [reply pkg/network/namescheme/networknamescheme.go]: Done.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/bef7ab877796052d651aea937ada7e99e24669e1..b6a99d8d397b6f0d230c3e471bc5277bbcdb36c5): Fixed @ormergi's [comment](https://github.com/kubevirt/kubevirt/pull/13130#discussion_r1819064956).

File: pkg/network/multus/status.go
Comment (ormergi) [reply pkg/network/multus/status.go:33]: Its not about approach, semantically this slice represent a pod network status, the fact that it has multiple interfaces, the slice as whole represent the pod network status.
In addition network-stautses reads bad, I think the term should be network-status because in all contexts we refer to this slice as the pod network status (as whole).

If you want to stick with referring to the slice as each interface status, I think the correct form is network**s**-status.

File: pkg/network/deviceinfo/deviceinfo_test.go
Comment (ormergi) [reply pkg/network/deviceinfo/deviceinfo_test.go:101]: This test is stable because the expected map has single entry, so nothing to do here, sorry for the noise.

File: pkg/network/pod/annotations/generator.go
Comment (ormergi) [reply pkg/network/pod/annotations/generator.go:117]: Thanks for explaining, I think we good.

File: pkg/virt-controller/watch/vmi/vmi.go
Comment (ormergi) [reply pkg/virt-controller/watch/vmi/vmi.go:2179]: I think it can happen in case Multus network-status annotation is maflormed/invalid, right?
Before the dynamic-nic-name PR changed, `GenerateCNIAnnotationFromNameScheme` would recive name-scheme map that is non-empty.
I suppose this is why we had no test for that, there was no scenario where the mapping is empty.

Comment (ormergi) [reply pkg/virt-controller/watch/vmi/vmi.go:2179]: Anyhow, it can be done on a follow-up PR because it didnt caused by this PR changed, sorry for the noise.

File: pkg/network/multus/status.go
Comment (orelmisan) [reply pkg/network/multus/status.go:33]: > Its not about approach I think the definition of the network-status slice is interpreted wrongly here.
Semantically this slice represent a pod network status, the fact that it has multiple interfaces, the slice as whole represent the pod network status.

How would you refer to each entry in case the whole slice is called `networkStatus`?

> In addition network-stautses reads bad, I think the term should be network-status because in all contexts we refer to this slice as the pod network status (as whole).

Following #13018 and this PR - it is referred to it as `networkStatuses` all over.

Comment (ormergi) [reply pkg/network/multus/status.go:33]: > How would you refer to each entry in case the whole slice is called networkStatus?

If you refer the whole slice as `networksStatus`, each element could be referred to `networkIfaceStatus`.
I also suggested to refer to the slice as `networksStatus` as an alternative, I think its the correct form (having the last adjective plural).

> Following https://github.com/kubevirt/kubevirt/pull/13018 and this PR - it is referred to as `networkStatuses` all over.

I am aware, I also commented in previous PRs, I am raising it again because it getting to other places.
Hopefully I will manage to convince you this time :slightly_smiling_face:

Comment (EdDev) [reply pkg/network/multus/status.go:33]: I do not understand the amount of investment in this thread.

You can convince me by ref the multus-cni source code or by being focused (avoiding lengthy explanations about the meaning).

My only request is to wrap this up quickly and try to prefer short over long names when possible:

net over network
iface over interface

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/b6a99d8d397b6f0d230c3e471bc5277bbcdb36c5..aab53ebcefc8dfd43ea7baed1c1157ca4006c20e): Rebase.

File: pkg/network/multus/status.go
Comment (ormergi) [reply pkg/network/multus/status.go:33]: Long story short I think the `networkStatuses` is wrong, correct form is `networksStatus`, please consider changing it instead of adding more of the former naming.

Comment (orelmisan) [reply pkg/network/multus/status.go:33]: From the Multus source code:
https://github.com/k8snetworkplumbingwg/multus-cni/blob/ecf5854ca9ce4c064ab4939808868747d7fa438f/vendor/github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/utils/net-attach-def.go#L49-L48

I've considered, and decided `networkStatuses` makes sense.
If there aren't additional issues with the PR, I suggest to move forward with it.
A follow-up PR could suggest renaming all the instances of this slice.

Comment (ormergi) [reply pkg/network/multus/status.go:33]: I am still think its not the right form but not blocking this PR on it

**Review by ormergi — APPROVED** [review]

File: pkg/network/multus/status.go
Comment (orelmisan) [reply pkg/network/multus/status.go:33]: Thank you @ormergi.

File: pkg/network/deviceinfo/deviceinfo.go
Comment (nirdothan) [inline pkg/network/deviceinfo/deviceinfo.go:35]: can't we get rid of this line?
we can replace `networkStatusEntry, exist := multusInterfaceNameToNetworkStatus[multusInterfaceName]` with a [binary search](https://pkg.go.dev/slices#BinarySearchFunc) over the `interfaces` slice

File: pkg/network/namescheme/networknamescheme.go
Comment (nirdothan) [inline pkg/network/namescheme/networknamescheme.go:129]: The function name is hiding what it creates. Please rename it and not to `CreateNetworkNameSchemeByPodNetworkStatus2` :)

**Review by nirdothan — COMMENTED** [review]
thanks @orelmisan - reviewed the first 3 commits and it looks good. see me comments

File: pkg/network/deviceinfo/deviceinfo.go
Comment (orelmisan) [reply pkg/network/deviceinfo/deviceinfo.go]: By the end of the PR, this function is used twice:
1. By the `pkg/network/deviceinfo` package.
2. By the `pkg/network/vmicontroller` package.

File: pkg/network/multus/status_test.go
Comment (orelmisan) [reply pkg/network/multus/status_test.go:37]: The unit test serves (at least) three purposes:
1. Testing the code's behavior.
2. Making sure the code's behavior doesn't change - thus guarding us from regression.
3. Documenting the code's behavior.

File: pkg/network/pod/annotations/generator.go
Comment (orelmisan) [reply pkg/network/pod/annotations/generator.go:108]: Following this PR, only `multus.NetworkStatusesFromPod` does the actual parsing / unmashalling.
In case there is an umarshaling error, there will be a log entry and a nil slice will be returned.

File: pkg/network/deviceinfo/deviceinfo.go
Comment (orelmisan) [reply pkg/network/deviceinfo/deviceinfo.go:35]: From the binarySearch docs:
> The slice must be sorted in increasing order, where "increasing" is defined by cmp

To the best of my knowledge, the `networkv1.NetworkStatus` slices are not ordered in a way that a binary search can utilize.
I encourage you to explore this subject in a follow-up.

File: pkg/network/namescheme/networknamescheme.go
Comment (orelmisan) [reply pkg/network/namescheme/networknamescheme.go:129]: When looking at an exported Go identifier, we need to remember that it is prefixed by its package name, so this specific function is actually called `namescheme.CreateFromNetworkStatuses` when called from other packages.
This is a recurring idiom in Go.

In case the name is still off, feel free to suggest another name.

You will be happy to see that one of the commits in this PR removed the `2` suffix from the function you were referring to.

File: pkg/network/deviceinfo/deviceinfo.go
Comment (nirdothan) [reply pkg/network/deviceinfo/deviceinfo.go:35]: right :( then we can use https://pkg.go.dev/slices#IndexFunc, this slice has got to be very small, so it doesn't really matter how we search through it

File: pkg/network/multus/status_test.go
Comment (nirdothan) [reply pkg/network/multus/status_test.go:37]: if we must... I'm generally not keen on writing long tests to trivial functions, but if the project policy is to do so then I can accept it

File: pkg/network/deviceinfo/deviceinfo.go
Comment (nirdothan) [reply pkg/network/deviceinfo/deviceinfo.go]: what I meant was that you might want to indicate in the commit that `mapMultusInterfaceNameToNetworkStatus` gets deleted  in a later commit

File: pkg/network/multus/status_test.go
Comment (EdDev) [reply pkg/network/multus/status_test.go:37]: I agree there is an unbalance relation between the production and its tests.

We could attempt to explore why:
- Some tests that have helped during development are not neseserry relevant once the work completed. E.g. if the production code does not differentiate between a single and multiple entries, then there is no point to keep individual tests for each.
- Extra data in tests may serve as a good context to understand what is happening, but in some cases can confuse the reader. E.g. is the `Default` field playing a role in the indexing? If not, why is it mentioned?
- Tests are focused on low level units, like functions, instead of higher level units that aim to provide a wider solution. I.e. if a test can cover the functionality at a higher abstraction level, it will usually be a better fit, especially when the implementation is trivial.

Looking back at the tests, they are lacking a critical characteristic of the indexing.
In case the index value is missing, we end up in a wired situation where there is one an empty string as an index that contains a random value (based on the last entry that has the pod iface name empty).

Making this exposed requires a more pedantic look at this, e.g. considering to ignore such entries and not index them.
Or as suggested above, depend on higher level unit tests to cover this helper.

File: pkg/network/deviceinfo/deviceinfo.go
Comment (EdDev) [reply pkg/network/deviceinfo/deviceinfo.go:35]: @nirdothan , I think it is how one defines elegant code.
If we go over a list and we search again and again through it, it becomes a N^2 cost.

While N is small, this does not really matter in terms of performance, but it does cause it to look "ugly".

So for the sake of an elegant code, we pass the whole list, index per what we need and then go over some loop and check existence. For most, this is seen as more elegant, but it may be just a subjective thing that we got used to. Personally I find searching again and again in the same list ugly, no matter the size of the list.

File: pkg/network/deviceinfo/deviceinfo_test.go
Comment (EdDev) [inline pkg/network/deviceinfo/deviceinfo_test.go:81]: nit: A `nil` may be more direct than an empty explicit list.
nit: You probably should have left lower-case `w`.

File: pkg/network/pod/annotations/generator.go
Comment (EdDev) [reply pkg/network/pod/annotations/generator.go:108]: @nirdothan , please feedback if you got convinced or not so this thread can continue or reach resolution.

**Review by EdDev — COMMENTED** [review]
Nice cleanup, thanks!

File: pkg/network/deviceinfo/deviceinfo_test.go
Comment (orelmisan) [reply pkg/network/deviceinfo/deviceinfo_test.go]: Done.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/aab53ebcefc8dfd43ea7baed1c1157ca4006c20e..c54460bbd6956a679245857168aa895d19170644): Fixed @EdDev's [comment](https://github.com/kubevirt/kubevirt/pull/13130#discussion_r1836747134).

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/c54460bbd6956a679245857168aa895d19170644..5eb5548d6a76bca4d615c931f0639f90b53a10eb): Rebase.

File: pkg/network/multus/status_test.go
Comment (orelmisan) [reply pkg/network/multus/status_test.go]: The unit tests for `NetworkStatusesByPodIfaceName` were simplified.
Added a comment to the first commit message discussing the issue of empty string as a key as a result of empty interface field.

File: pkg/network/deviceinfo/deviceinfo.go
Comment (EdDev) [reply pkg/network/deviceinfo/deviceinfo.go:35]: > I'm not so happy about making another traversal

It is one traversal vs N traversals, so I do not know why you see it as "making another".
That is exactly why I think it is ugly and not elegant.

> and building yet another map just to later check if an element exists in a slice.

Yes, that is the cost, to convert one data structure to another in order to efficiently lookup entries.
That new data structure has a memory cost, but as you pointed out, is negligible due to the size.

> in HL programming languages such as Python or Java you would never do that.

That is because those languages usually have dedicated data structures which are optimized for searching in the first place.

> I could research multiple solutions that are more elegant, maybe by changing the type/s returned from `CreateFromNetworkStatuses()`.

At its base, that means you either can avoid the search, use a better data structure than slice, etc.

File: pkg/virt-controller/watch/vmi/vmi.go
Comment (EdDev) [inline pkg/virt-controller/watch/vmi/vmi.go:1]: @orelmisan , you will need to get a root approval ack on this file.

**Comment by fossedihelm** [comment]
/approve
for controller changes

**Pull request merged**
